docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,504 @@
+defined in terms of still more algorithms. It can be intimidating to consider
+exhaustively testing one of those "nested" algorithms, especially when they are
+shared by many different APIs.
+
+In general, you should start by writing only "surface tests" for the nested
+algorithms. That means only verifying that they exhibit the basic behavior you
+are expecting.
+
+It's definitely important to test exhaustively, but it's just as important to
+do so in a structured way. Reach out to the test suite's maintainers to learn
+if and how they have already tested those algorithms. In many cases, it's
+acceptable to test them in just one place (and maybe through a different API
+entirely), and rely only on surface-level testing everywhere else. While it's
+always possible for more tests to uncover new bugs, the chances may be slim.
+The time we spend writing tests is highly valuable, so we have to be efficient!
+
+*Example* The following algorithm from [the DOM
+standard](https://dom.spec.whatwg.org/) powers
+[`document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector):
+
+> To **scope-match a selectors string** *selectors* against a *node*, run these
+> steps:
+>
+> 1. Let *s* be the result of [parse a
+>    selector](https://drafts.csswg.org/selectors-4/#parse-a-selector)
+>    *selectors*.
+>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4)
+> 2. If *s* is failure, then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
+>    "[`SyntaxError`](https://heycam.github.io/webidl/#syntaxerror)"
+>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
+> 3. Return the result of [match a selector against a
+>    tree](https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree)
+>    with *s* and *node*'s
+>    [root](https://dom.spec.whatwg.org/#concept-tree-root) using [scoping
+>    root](https://drafts.csswg.org/selectors-4/#scoping-root) *node*.
+>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4).
+
+As described earlier in this guide, we'd certainly want to test the branch
+regarding the parsing failure. However, there are many ways a string might fail
+to parse--should we verify them all in the tests for `document.querySelector`?
+What about `document.querySelectorAll`? Should we test them all there, too?
+
+The answers depend on the current state of the test suite: whether or not tests
+for selector parsing exist and where they are located. That's why it's best to
+confer with the people who are maintaining the tests.
+
+### Avoid excessive breadth
+
+Many algorithms operate on input that takes the form of a scalar value. When
+the set of values is finite, it can be tempting to test them all exhaustively.
+When the set is very large, test authors can reduce repetition by defining
+tests programmatically in loops.
+
+Using advanced control flow techniques to dynamically generate tests has a
```

Comment:
Does this refer to nested for loops? I disagree that this makes tests harder to understand, often it's clearer than if the same test is copy-pasted over and over. But too many subtests and time to run the tests is a problem.

I know that @domenic and @annevk also have opinions on this and especially @domenic might be more skeptical of generated tests than I am.

Reply:
I am quite skeptical of generated tests. The `web-animations` test suite is *amazing*, but it uses a lot of generated tests and they cause two problems:

i. Hard to read quickly (see https://github.com/web-platform-tests/wpt/blob/master/web-animations/animation-model/animation-types/property-types.js or things like https://github.com/web-platform-tests/wpt/blob/master/web-animations/interfaces/KeyframeEffect/getKeyframes.html),

ii. Maybe worse, really hard to debug quickly. Often pulling out a single failing tests means first finding the underlying 'data-like' file, then figuring out how to comment out the parts you don't need, etc etc.

I think generated tests have value, but are a dangerous tool and can easily stray into being used for their own sake.

Reply:
A counter example with nested for loops (`forEach` actually) that I think is *just fine*:
https://github.com/web-platform-tests/wpt/blob/master/fullscreen/api/historical.html

Unwinding these tests would not be an improvement at all, IMHO.

So, rather than pointing to "control flow techniques" as generally having drawbacks, I'd point out how it can go wrong, while being neutral on it in general or even endorsing the sensible uses.

Reply:
> I think it's easier to understand the problem by being more direct: When generating tests, it's easy to end up with a lot of tests which introduces the problem that running the tests takes too much time.

@zcorpan That framing doesn't account for the objection I described above. Readers may say, "of course more tests take more time; if we were optimizing for time, we wouldn't run any tests at all!" That's why I think whatever wording we choose, we have to acknowledge the "relative value" of the tests in some way.

> A counter example with nested for loops (`forEach` actually) that I think is *just fine*:
> https://github.com/web-platform-tests/wpt/blob/master/fullscreen/api/historical.html
>
> Unwinding these tests would not be an improvement at all, IMHO.

Loops may be a red herring. I get the most anxious when I see assertions or subtests inside of `if` statements. In [the DOMTokenList test I referenced earlier](https://github.com/web-platform-tests/wpt/blob/fee6738f9523f5e257accfaea8ddfef2e28cfbcd/dom/lists/DOMTokenList-coverage-for-attributes.html) the `testAttr` function is critical, and understanding what is being tested requires performing a mental transformation that reminds me of linear algebra.

But now I'm just enjoying the philosophical discussion. I don't think we should wrestle with that in this guide.

> So, rather than pointing to "control flow techniques" as generally having drawbacks, I'd point out how it can go wrong, while being neutral on it in general or even endorsing the sensible uses.

Sounds good. We were almost there, so I just softened the wording a bit. Does the latest iteration (no apologies for the pun) look good to you?

Reply:
Those changes are in https://github.com/web-platform-tests/wpt/pull/19628/commits/f8930999036024eca2c51e848dbfefb122be167b and while it's not quite as kind to generated tests as I would be, I recognize that others have a worse experience with them, and the wording is an OK middle ground.

Reply:
I agree with @zcorpan in https://github.com/web-platform-tests/wpt/pull/19628/commits/10ae79576bcbc447bedaf7bb458537d4ac109e96#r337517178, but will start a new thread about that specifically, as this thread will be shown as outdated by GitHub and hard to find.